8346954: [JMH] jdk.incubator.vector.MaskedLogicOpts fails due to IndexOutOfBoundsException

xyyNicole opened the pull request:
Suite `MaskedLogicOpts.maskedLogicOperationsLong512()` failed on both x86 and AArch64 with the following error:

```
java.lang.IndexOutOfBoundsException: Index 252 out of bounds for length 249
```

The variable `long256_arr_idx` is misused when indexing `LongVector l2`, `l3`, `l4`, `l5` in function `maskedLogicOperationsLongKernel()` resulting in the IndexOutOfBoundsException error. On the other hand, the unified index for 128-bit, 256-bit and 512-bit species might not be proper since it leaves gaps in between when accessing the data for 128-bit and 256-bit species. This will unnecessarily include the noise due to cache misses or (on some targets) prefetching additional cache lines which are not usable, thereby impacting the crispness of microbenchmark.

Hence, we improved the benchmark from several aspects,
1. Used sufficient number of predicated operations within the vector loop while minimizing the noise due to memory operations.
2. Modified the index computation logic which can now withstand any ARRAYLEN without resulting in an IOOBE.
3. Removed redundant vector read/writes to instance fields, thus eliminating significant boxing penalty which translates into throughput gains.

<!-- Anything below this marker will be automatically updated, please do not edit manually! -->
---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8346954](https://bugs.openjdk.org/browse/JDK-8346954): [JMH] jdk.incubator.vector.MaskedLogicOpts fails due to IndexOutOfBoundsException (**Bug** - P4)


### Reviewers
 * [Xiaohong Gong](https://openjdk.org/census#xgong) (@XiaohongGong - Committer) Review applies to [083bedec](https://git.openjdk.org/jdk/pull/22963/files/083bedec04d5ab78a420e156e74c1257ce30aee8)
 * [Jatin Bhateja](https://openjdk.org/census#jbhateja) (@jatin-bhateja - **Reviewer**)

### Contributors
 * Jatin Bhateja `<jbhateja@openjdk.org>`

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk.git pull/22963/head:pull/22963` \
`$ git checkout pull/22963`

Update a local copy of the PR: \
`$ git checkout pull/22963` \
`$ git pull https://git.openjdk.org/jdk.git pull/22963/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 22963`

View PR using the GUI difftool: \
`$ git pr show -t 22963`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/22963.diff">https://git.openjdk.org/jdk/pull/22963.diff</a>

</details>
<details><summary>Using Webrev</summary>

[Link to Webrev Comment](https://git.openjdk.org/jdk/pull/22963#issuecomment-2638273429)
</details>

xyyNicole commented:
/covered

xyyNicole commented:
I am a member of Nvidia Java compiler team.

xyyNicole commented:
> I am a member of Nvidia Java compiler team.

BTW, Nvidia has signed the OCA recently. Please help to check. Thanks.

XiaohongGong commented:
/label hotspot-compiler

eme64 commented:
@jatin-bhateja Could you have a look at these changes? You wrote the test originally.

eme64 commented:
@xyyNicole I see this has been in OCA-verify mode for 2 weeks. I reached out internally and hope this can go through soon.

eme64 commented:
Oh, the OCA-verify is still stuck. I'm sorry about that :see_no_evil: 
I pinged my manager @TobiHartmann , he will reach out to see what's the issue.

xyyNicole commented:
Thanks @eme64. The OCA finally passed! 😄

xyyNicole commented:
Hi @jatin-bhateja Would you help to review the patch and verify the changes? Thanks.

XiaohongGong reviewed (approved):
Looks good to me. Thanks for the fixing!

xyyNicole commented:
Hi @eme64, do you see any risks here? Would you please help to review the patch? Thanks.

jatin-bhateja commented on test/micro/org/openjdk/bench/jdk/incubator/vector/MaskedLogicOpts.java:
```diff
@@ -126,10 +120,7 @@ public void init() {
     @Setup(Level.Invocation)
     public void init_per_invoc() {
         int512_arr_idx = (int512_arr_idx + 16) & (ARRAYLEN-1);
```
Benchmark assumes that ARRAYLEN is a POT value, thus it will also be good to use the % operator here for rounding, it will be expensive but will not impact the performance of the Benchmarking kernerls.

jatin-bhateja replied:
Please try with following command line
`java -jar target/benchmarks.jar -f 1 -i 2 -wi 1 -w 30 -p ARRAYLEN=30 MaskedLogic`

xyyNicole replied:
Thanks for pointing that out. Typically, ARRAYLEN is almost always a POT value, which is also assumed by many other benchmarks. Are we realistically going to test with an ARRAYLEN of 30? 

I think the POT assumption is reasonable for our purposes.

PaulSandoz replied:
It's a reasonable assumption. Since `ARRAYLEN` is a parameter of the benchmark we should enforce that constraint in benchmark initialization method, checking if the value is POT and failing otherwise.

xyyNicole replied:
Hi, @PaulSandoz, thanks for your suggestions.

In @jatin-bhateja's latest updates, non-POT `ARRAYLEN` values have been supported by processing only the elements up to the nearest multiple of the vector length. In that case, we can use a wider range of `ARRAYLEN` values while focusing on the main loop for vector functionality.

jatin-bhateja commented on test/micro/org/openjdk/bench/jdk/incubator/vector/MaskedLogicOpts.java:
```diff
@@ -243,10 +234,10 @@ public void bitwiseBlendOperationInt128() {
     @CompilerControl(CompilerControl.Mode.INLINE)
```
Benchmarking kernels are forced inlines, so passing a specific index value may help.

xyyNicole commented on test/micro/org/openjdk/bench/jdk/incubator/vector/MaskedLogicOpts.java:
```diff
@@ -126,10 +120,7 @@ public void init() {
-        long256_arr_idx = (long256_arr_idx + 4) & (ARRAYLEN-1);
     }
 
     @CompilerControl(CompilerControl.Mode.INLINE)
```
Thanks, @jatin-bhateja, for your thorough review.

Yes, you're right, the current design does introduce gaps when accessing the data. And other cases in this suite are also designed in a similar way. I am wondering if you're interested in refactoring the code to address this more comprehensively.

The primary goal of this pull request is to address an out-of-bounds issue that was blocking our tests. We aimed to ensure all test cases pass and become unblocked. Additionally, our available hardware resources limit our ability to rigorously test a wide range of scenarios at this time.

Therefore, we opted to maintain consistency with the existing logic in other cases within the suite and simply fix the crash issue. This approach allows us to unblock the current tests and keep things moving for now.

eme64 commented:
@xyyNicole @jatin-bhateja I think it is reasonable to just fix the benchmark so that it still has the same behaviour, just without the out-of-bounds exception. @jatin-bhateja you originally wrote the benchmark, and it could make sense if you fixed it up to what it should be more ideally. @xyyNicole I propose that we file a follow-up RFE to fix the benchmark, and just mention that issue in the benchmark.

What do you think?

jatin-bhateja commented:
> @xyyNicole @jatin-bhateja I think it is reasonable to just fix the benchmark so that it still has the same behaviour, just without the out-of-bounds exception. @jatin-bhateja you originally wrote the benchmark, and it could make sense if you fixed it up to what it should be more ideally. @xyyNicole I propose that we file a follow-up RFE to fix the benchmark, and just mention that issue in the benchmark.
> 
> What do you think?

Hi @xyyNicole , 

May I request you to kindly file a follow up RFE mentioning the discussed issues.
Performance of lanewise operations works best without any noise if its fed contiguous memory location.

Best Regards,
Jatin

xyyNicole commented:
> > @xyyNicole @jatin-bhateja I think it is reasonable to just fix the benchmark so that it still has the same behaviour, just without the out-of-bounds exception. @jatin-bhateja you originally wrote the benchmark, and it could make sense if you fixed it up to what it should be more ideally. @xyyNicole I propose that we file a follow-up RFE to fix the benchmark, and just mention that issue in the benchmark.
> > What do you think?
> 
> Hi @xyyNicole ,
> 
> May I request you to kindly file a follow up RFE mentioning the discussed issues. Performance of lanewise operations works best without any noise if its fed contiguous memory locations.
> 
> Best Regards, Jatin

xyyNicole commented:
Sure. Since I am very new to openJDK, I asked my teammate for help to file the follow-up RFE.

Here is the https://bugs.openjdk.org/browse/JDK-8350215 with description of the discussed issues.

jatin-bhateja commented:
> Sure. Since I am very new to openJDK, I asked my teammate for help to file the follow-up RFE.
> 
> Here is the https://bugs.openjdk.org/browse/JDK-8350215 with description of the discussed issues.

Hi @xyyNicole , 

I have modified the benchmark keeping its essence intact, i.e. to use sufficient number of predicated operations within the vector loop while minimizing the noise due to memory operations.
Modified the index computation logic which can now withstand any ARRAYLEN without resulting in an IOOBE.
Removed redundant vector read/writes to instance fields, thus eliminating significant boxing penalty which translates into throughput gains.

Please feel free to include it along with this patch.
[MaskedLogicOpts.txt](https://github.com/user-attachments/files/18882179/MaskedLogicOpts.txt)

jatin-bhateja commented:
> Sure. Since I am very new to openJDK, I asked my teammate for help to file the follow-up RFE.
> 
> Here is the https://bugs.openjdk.org/browse/JDK-8350215 with description of the discussed issues.

Hi @xyyNicole , 

I have modified the benchmark keeping its essence intact, i.e. to use sufficient number of predicated operations within the vector loop while minimizing the noise due to memory operations.
Modified the index computation logic which can now withstand any ARRAYLEN without resulting in an IOOBE.
Removed redundant vector read/writes to instance fields, thus eliminating significant boxing penalty which translates into throughput gains.

Please feel free to include it along with this patch.
[MaskedLogicOpts.txt](https://github.com/user-attachments/files/18884093/MaskedLogicOpts.txt)

jatin-bhateja reviewed (approved):
LGTM

xyyNicole commented:
> > Sure. Since I am very new to openJDK, I asked my teammate for help to file the follow-up RFE.
> > Here is the https://bugs.openjdk.org/browse/JDK-8350215 with description of the discussed issues.
> 
> Hi @xyyNicole ,
> 
> I have modified the benchmark keeping its essence intact, i.e. to use sufficient number of predicated operations within the vector loop while minimizing the noise due to memory operations. Modified the index computation logic which can now withstand any ARRAYLEN without resulting in an IOOBE. Removed redundant vector read/writes to instance fields, thus eliminating significant boxing penalty which translates into throughput gains.
> 
> Please feel free to include it along with this patch. [MaskedLogicOpts.txt](https://github.com/user-attachments/files/18884093/MaskedLogicOpts.txt)

Hi @jatin-bhateja , 

Thank you for your contributions. I've incorporated your modifications into this patch and made some minor formatting adjustments. 😄

xyyNicole commented:
/contributor add @jatin-bhateja

xyyNicole commented:
Hi, @eme64, do you have any additional comments for this patch? Is it ready to be merged?

XiaohongGong commented:
Still looks good to me!

xyyNicole commented:
@XiaohongGong @eme64 @jatin-bhateja @PaulSandoz  Thanks for your review. I'm going to integrate the patch.

/integrate

XiaohongGong commented:
/sponsor
